doc/sphinx/src/contributing.rst

```diff
@@ -2,3 +2,120 @@
 
 Contributing
 =============
+
+Useful code structure information
+---------------------------------
+
+The ``Variant`` class and ``mpark::variant``
+````````````````````````````````````````````
+
+Work in progress. Things to cover:
+
+*  Type erasure
+*  Unified API
+
+The CRTP slass structure and static polymorphism
+````````````````````````````````````````````````
+
+Each of the EOS models in ``singularity-eos`` inherits from a base class in
+order to centralize default functionality and avoid code duplication. The two
+main examples of this are the vector overloads and the ``PTofRE`` scalar lookup
+function. In the vector overloads, a simple for loop is used to loop over the
+set of states provided to the function and then call the scalar version on each
+state. The ``PTofRE`` function is designed to provide a common method for
+getting the needing information for a PTE solve from an EOS. Both of these
+features are not dependent on the specific EOS for their definition, but in the
+case of the vector overloads, they *do* need to access methods in the derived
+class.
+
+The vector overloads in the base class take the following form (in pseudocode):
+
+.. code-block:: c++
+
+    template <typename RealIndexer, typename ConstRealIndexer, typename LambdaIndexer>
+    inline void
+    TemperatureFromDensityInternalEnergy(ConstRealIndexer &&rhos, ConstRealIndexer &&sies,
+                                         RealIndexer &&temperatures, const int num,
+                                         LambdaIndexer &&lambdas) const {
+    for (int i = 0; i < num; i++) {
+        temperatures[i] = eos.TemperatureFromDensityInternalEnergy(rhos[i],
+            sies[i], lambdas[i])
+
+where the base class basically needs to call the implementation of the scalar
+lookup in the specific EOS. However, this means that the base class needs to
+have knowledge of which class is being derived from it in order to call the
+correct EOS implementation.
+
+The challenge in making this sort of code performance portable is that on a CPU
+this sort of type deduction (i.e. what specific EOS implementaiton to call)
+would be done at runtime through vtables. In a heterogenous computing enviroment
+though, the vtable may point to member functions that do not exist on the
+device, causing a segfault. We could have used a similar technique to the
+modifier classes and pass the EOS as a template paramter, but then the vector
+function calls could only be achieved by creating vector modifiers of all the
+implemented EOS.
```

Comment:
The issue isn't really segfaults. Runtime inheritance is possible on GPU, just cumbersome and less performant. I would rephrase this to something like:

> The standard solution to this problem would be "run-time inheritence," where type deduction is performed at run-time. While this is possible on GPU, it becomes cumbersome, as the user must be very explicit about class inheritence. Moreover, run-time inheritence relies on relocatable device code, which is not as performant on device, thanks to weaker cross-compilation unit optimization. (We note that to fully leverage relocatable device code, the entire library must be made to be header-only.) 
>
>  We could have used a similar technique to the
modifier classes and pass the EOS as a template paramter, but then the vector
function calls could only be achieved by creating vector modifiers of all the
implemented EOS.

Reply:
"We note that to fully leverage relocatable device code (RDC), the entire library must be made header-only" -> "We note that to obtain full performance on device and to build with compilers that don't support (RDC), the entire library must be made header-only"